### DISCUSSION TITLE
Noob here! How to change key mappings for nvim-cmp

### DISCUSSION THREAD

**Discussion by fabricetriboix** [NvChad/NvChad#2832] [Q&A] [answered]
Hello,

When the autocompletion menu pops up, I want to be able to select the next entry with the "down" key instead of "tab", and the previous entry with the "up" key instead of "shift+tab". 

I spent a couple of hours searching the internet, but I can't find anything... If anybody could help me, that would be great!

I just installed everything a couple of days ago from scratch, so everything is pretty much the latest stable version available, with almost no customisation.

Thanks!

**Comment by fabricetriboix** [comment]
I had a look around in my nvim files, and I could see this in the `/.local/share/nvim/lazy/nvim-cmp/lua/cmp/config/mapping.lua` file:

```lua
mapping.preset.insert = function(override)
  return merge_keymaps(override or {}, {
    ['<Down>'] = {
      i = mapping.select_next_item({ behavior = types.cmp.SelectBehavior.Select }),
    },
    ['<Up>'] = {
      i = mapping.select_prev_item({ behavior = types.cmp.SelectBehavior.Select }),
    },
    ['<C-n>'] = {
      i = function()
        local cmp = require('cmp')
        if cmp.visible() then
          cmp.select_next_item({ behavior = types.cmp.SelectBehavior.Insert })
        else
          cmp.complete()
        end
      end,
    },
    ['<C-p>'] = {
      i = function()
        local cmp = require('cmp')
        if cmp.visible() then
          cmp.select_prev_item({ behavior = types.cmp.SelectBehavior.Insert })
        else
          cmp.complete()
        end
      end,
    },
    ['<C-y>'] = {
      i = mapping.confirm({ select = false }),
    },
    ['<C-e>'] = {
      i = mapping.abort(),
    },
  })
end
```

Am I correct I thinking that this code enforce `C-n` and `C-p` to navigate the popup menu where I wanted up and down keys? If yes, is it possible to overwrite this by configuring the `nvim-cmp` plugin?

Thanks!

**Reply by siduck** [reply]
@fabricetriboix i'll give u a hint, look at the last example https://nvchad.com/docs/config/plugins , now implement in for CMP , first look  our default cmp config in nvchad's main repo. 1) import it  2) modify it 3) return it! 

all of that in opts function

**Reply by fabricetriboix** [reply]
All right, thanks @siduck , I'll try. Would you mind pointing me to the NvChad default config for nvim-cmp? Thanks!

**Reply by siduck** [reply]
@fabricetriboix  https://github.com/NvChad/NvChad/blob/v2.5/lua/nvchad/configs/cmp.lua

**Reply by fabricetriboix** [reply]
Thanks @siduck !

**Reply by fabricetriboix** [reply]
@siduck It worked! First I copied the entire code and modified what I need, then I tried to `require` the default NvChad config and overwrite what I wanted and that worked. So here is a snippet of my `init.lua` file:

```lua
  {
    "hrsh7th/nvim-cmp",
    opts = function()
      local cmp = require("cmp")
      local conf = require("nvchad.configs.cmp")

      conf.mapping = {
        ["<Up>"] = cmp.mapping.select_prev_item(),
        ["<Down>"] = cmp.mapping.select_next_item(),
        ["<Tab>"] = cmp.mapping.confirm {
          behavior = cmp.ConfirmBehavior.Replace,
          select = true,
        },
      }
      return conf
    end,
  },
```

Thanks for your help!

**Reply by siduck** [reply]
@fabricetriboix  but i think your method is slightly wrong,  you're assigning mapping table your own table which just has 3 keys!  which means that you'll be losing all  of our default cmp mappings, which also has luasnip stuff 

you could do this, `conf.mapping["<Up>"]` or     assign your mapping table to `local mymappings` variable and then merge it with conf.mapping, like 
```lua
conf.mapping = vim.tbl_deep_extend("force", conf.mapping, mymappings)
```

"force" there means that the changes from your table gets merged to the original table i.e conf.mapping

**Reply by fabricetriboix** [reply, marked as answer]
Oh, I see! Thanks @siduck . So the final solution is:

```lua
  {
    "hrsh7th/nvim-cmp",
    opts = function()
      local cmp = require("cmp")
      local conf = require("nvchad.configs.cmp")

      local mymappings = {
        ["<Up>"] = cmp.mapping.select_prev_item(),
        ["<Down>"] = cmp.mapping.select_next_item(),
        ["<Tab>"] = cmp.mapping.confirm {
          behavior = cmp.ConfirmBehavior.Replace,
          select = true,
        },
      }
      conf.mapping = vim.tbl_deep_extend("force", conf.mapping, mymappings)
      return conf
    end,
  },
```

**Reply by siduck** [reply]
@fabricetriboix  since the config function of cmp has only 1 line of code, you can do this too :grimacing:  as it doesnt need importing default cmp config! because the opts arg itself is handled by lazy and it'll give us default cmp config

```lua
{
	"hrsh7th/nvim-cmp",
	config = function(_, opts)
		local cmp = require("cmp")

		local mymappings = {
			["<Up>"] = cmp.mapping.select_prev_item(),
			["<Down>"] = cmp.mapping.select_next_item(),
			["<Tab>"] = cmp.mapping.confirm({
				behavior = cmp.ConfirmBehavior.Replace,
				select = true,
			}),
		}

		opts.mapping = vim.tbl_deep_extend("force", opts.mapping, mymappings)
		cmp.setup(opts)
	end,
}
```
